File: api/src/icarus/routers/investigation.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import JSONResponse
from neo4j import AsyncSession

from icarus.dependencies import get_session
from icarus.models.investigation import (
    Annotation,
    AnnotationCreate,
    InvestigationCreate,
    InvestigationListResponse,
    InvestigationResponse,
    InvestigationUpdate,
    Tag,
    TagCreate,
)
from icarus.services import investigation_service as svc

router = APIRouter(tags=["investigations"])


@router.post(
    "/api/v1/investigations/",
    response_model=InvestigationResponse,
    status_code=201,
)
async def create_investigation(
    body: InvestigationCreate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> InvestigationResponse:
    return await svc.create_investigation(session, body.title, body.description)


@router.get("/api/v1/investigations/", response_model=InvestigationListResponse)
async def list_investigations(
    session: Annotated[AsyncSession, Depends(get_session)],
    page: Annotated[int, Query(ge=1)] = 1,
    size: Annotated[int, Query(ge=1, le=100)] = 20,
) -> InvestigationListResponse:
    investigations, total = await svc.list_investigations(session, page, size)
    return InvestigationListResponse(investigations=investigations, total=total)


@router.get(
    "/api/v1/investigations/{investigation_id}",
    response_model=InvestigationResponse,
)
async def get_investigation(
    investigation_id: str,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> InvestigationResponse:
    result = await svc.get_investigation(session, investigation_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Investigation not found")
    return result


@router.put(
    "/api/v1/investigations/{investigation_id}",
    response_model=InvestigationResponse,
)
async def update_investigation(
    investigation_id: str,
    body: InvestigationUpdate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> InvestigationResponse:
    result = await svc.update_investigation(
        session, investigation_id, body.title, body.description
    )
    if result is None:
        raise HTTPException(status_code=404, detail="Investigation not found")
    return result


@router.delete(
    "/api/v1/investigations/{investigation_id}",
    status_code=204,
)
async def delete_investigation(
    investigation_id: str,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> None:
    deleted = await svc.delete_investigation(session, investigation_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Investigation not found")


@router.post(
    "/api/v1/investigations/{investigation_id}/entities/{entity_id}",
    status_code=201,
)
async def add_entity(
    investigation_id: str,
    entity_id: str,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> dict[str, str]:
    added = await svc.add_entity_to_investigation(session, investigation_id, entity_id)
    if not added:
        raise HTTPException(status_code=404, detail="Investigation or entity not found")
    return {"investigation_id": investigation_id, "entity_id": entity_id}


@router.post(
    "/api/v1/investigations/{investigation_id}/annotations",
    response_model=Annotation,
    status_code=201,
)
async def create_annotation(
    investigation_id: str,
    body: AnnotationCreate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> Annotation:
    return await svc.create_annotation(
        session, investigation_id, body.entity_id, body.text
    )


@router.get(
    "/api/v1/investigations/{investigation_id}/annotations",
    response_model=list[Annotation],
)
async def list_annotations(
    investigation_id: str,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[Annotation]:
    return await svc.list_annotations(session, investigation_id)


@router.post(
    "/api/v1/investigations/{investigation_id}/tags",
    response_model=Tag,
    status_code=201,
)
async def create_tag(
    investigation_id: str,
    body: TagCreate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> Tag:
    return await svc.create_tag(session, investigation_id, body.name, body.color)


@router.get(
    "/api/v1/investigations/{investigation_id}/tags",
    response_model=list[Tag],
)
async def list_tags(
    investigation_id: str,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[Tag]:
    return await svc.list_tags(session, investigation_id)


@router.post(
    "/api/v1/investigations/{investigation_id}/share",
    response_model=dict[str, str],
)
async def generate_share_link(
    investigation_id: str,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> dict[str, str]:
    token = await svc.generate_share_token(session, investigation_id)
    if token is None:
        raise HTTPException(status_code=404, detail="Investigation not found")
    return {"share_token": token}


@router.get("/api/v1/shared/{token}", response_model=InvestigationResponse)
async def get_shared_investigation(
    token: str,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> InvestigationResponse:
    result = await svc.get_by_share_token(session, token)
    if result is None:
        raise HTTPException(status_code=404, detail="Shared investigation not found")
    return result


@router.get("/api/v1/investigations/{investigation_id}/export")
async def export_investigation(
    investigation_id: str,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> JSONResponse:
    investigation = await svc.get_investigation(session, investigation_id)
    if investigation is None:
        raise HTTPException(status_code=404, detail="Investigation not found")

    annotations = await svc.list_annotations(session, investigation_id)
    tags = await svc.list_tags(session, investigation_id)

    export_data = {
        "investigation": investigation.model_dump(),
        "annotations": [a.model_dump() for a in annotations],
        "tags": [t.model_dump() for t in tags],
    }
    return JSONResponse(content=export_data)

File: frontend/src/stores/investigation.ts
import { create } from "zustand";

import type { Annotation, Investigation, Tag } from "@/api/client";
import * as api from "@/api/client";

interface InvestigationState {
  // State
  activeInvestigationId: string | null;
  investigations: Investigation[];
  annotations: Annotation[];
  tags: Tag[];
  loading: boolean;
  error: string | null;

  // Actions
  setActiveInvestigation: (id: string | null) => void;
  fetchInvestigations: () => Promise<void>;
  createInvestigation: (title: string, description?: string) => Promise<Investigation>;
  deleteInvestigation: (id: string) => Promise<void>;
  updateInvestigation: (id: string, data: { title?: string; description?: string }) => Promise<void>;
  addEntity: (investigationId: string, entityId: string) => Promise<void>;
  fetchAnnotations: (investigationId: string) => Promise<void>;
  addAnnotation: (investigationId: string, entityId: string, text: string) => Promise<void>;
  fetchTags: (investigationId: string) => Promise<void>;
  addTag: (investigationId: string, name: string, color?: string) => Promise<void>;
}

export const useInvestigationStore = create<InvestigationState>((set, get) => ({
  activeInvestigationId: null,
  investigations: [],
  annotations: [],
  tags: [],
  loading: false,
  error: null,

  setActiveInvestigation: (id) => set({ activeInvestigationId: id }),

  fetchInvestigations: async () => {
    set({ loading: true, error: null });
    try {
      const res = await api.listInvestigations();
      set({ investigations: res.investigations, loading: false });
    } catch {
      set({ error: "Failed to load investigations", loading: false });
    }
  },

  createInvestigation: async (title, description) => {
    const inv = await api.createInvestigation(title, description);
    set((s) => ({ investigations: [inv, ...s.investigations] }));
    return inv;
  },

  deleteInvestigation: async (id) => {
    await api.deleteInvestigation(id);
    const state = get();
    set({
      investigations: state.investigations.filter((i) => i.id !== id),
      activeInvestigationId:
        state.activeInvestigationId === id ? null : state.activeInvestigationId,
    });
  },

  updateInvestigation: async (id, data) => {
    const updated = await api.updateInvestigation(id, data);
    set((s) => ({
      investigations: s.investigations.map((i) => (i.id === id ? updated : i)),
    }));
  },

  addEntity: async (investigationId, entityId) => {
    await api.addEntityToInvestigation(investigationId, entityId);
    // Re-fetch investigation to get updated entity_ids
    const updated = await api.getInvestigation(investigationId);
    set((s) => ({
      investigations: s.investigations.map((i) =>
        i.id === investigationId ? updated : i,
      ),
    }));
  },

  fetchAnnotations: async (investigationId) => {
    try {
      const annotations = await api.listAnnotations(investigationId);
      set({ annotations });
    } catch {
      set({ annotations: [] });
    }
  },

  addAnnotation: async (investigationId, entityId, text) => {
    const annotation = await api.createAnnotation(investigationId, entityId, text);
    set((s) => ({ annotations: [...s.annotations, annotation] }));
  },

  fetchTags: async (investigationId) => {
    try {
      const tags = await api.listTags(investigationId);
      set({ tags });
    } catch {
      set({ tags: [] });
    }
  },

  addTag: async (investigationId, name, color) => {
    const tag = await api.createTag(investigationId, name, color);
    set((s) => ({ tags: [...s.tags, tag] }));
  },
}));
